### ISSUE TITLE
Smoothness overlay

### ISSUE THREAD

**Issue by michalgwo** [streetcomplete/StreetComplete#5486]
### General
Affected tag(s) to be modified/added: [smoothness](https://wiki.openstreetmap.org/wiki/Key:smoothness)
Question asked: What’s the surface quality here?

We already have a quest asking for highway smoothness, so why not make it an overlay?

### Checklist
Checklist for quest suggestions (see [guidelines](https://github.com/streetcomplete/StreetComplete/blob/master/QUEST_GUIDELINES.md)):
- [x] 🚧 To be added tag is established and has a useful purpose
- [x] 🤔 Any answer the user can give must have an equivalent tagging (Quest should not reappear to other users when solved by one)
- [x] 🐿️ Easily answerable by any pedestrian from the outside but a survey is necessary
- [x] 💤 Not an overwhelming percentage of quests have the same answer (No spam)
- [x] 🕓 Applies to a reasonable number of map data (Worth the effort)

**Comment by westnordost** [comment]
That makes sense, actually, the quest should be _replaced_ by an overlay.

**Comment by rhhsm** [comment]
> That makes sense, actually, the quest should be _replaced_ by an overlay.

This would enable comparison between highways to be tagged and ones already tagged, which might lead some mappers to be tempted to highlight differences ("this road is worse than that one") while both sections actually fall within the same category (this road is at the bad end of `intermediate` and that one is at the good end of `intermediate`). I have my doubts if that is a good thing.

Anyway, for an impression of what such an overlay may look like, see here http://www.roads-bg.eu/#13/42.6721/23.3336 (green=`excellent` or `good`, orange=`intermediate`, red=`bad`, dark red=`very_bad`, only motor traffic roads are shown). It also shows my main area of SC survey activity :)

**Comment by matkoniecz** [comment]
> This would enable comparison between highways to be tagged and ones already tagged, which might lead some mappers to be tempted to highlight differences ("this road is worse than that one") while both sections actually fall within the same category (this road is at the bad end of intermediate and that one is at the good end of intermediate). I have my doubts if that is a good thing.

It would also encourage greater consistency, at least at local scale and make easier to spot already present wrong data.

**Comment by rhhsm** [comment]
> It would also encourage greater consistency, at least at local scale and make easier to spot already present wrong data.

Yes, that's true: I already used ivanatora's website above to find wrong data. The advantages of a smoothness overlay are probably bigger than the disadvantage I sketched above.

**Comment by mnalis** [comment]
> That makes sense, actually, the quest should be replaced by an overlay.

**Comment by mnalis** [comment]
> That makes sense, actually, the quest should be _replaced_ by an overlay.

While I'd like to have smoothness overlay and would use it sometimes, I must note that I quite **_dislike_** the trend of **removing quests in favor or overlays**. I even (mostly) agree that the quest should be auto-disabled when overlay is active, and  understand that _"less duplication of similar code, the better"_ as a general rule, but **not when it comes at heavy price for usability**. I'd certainly use smoothness overlay much less than smoothness quest.


I must say I personally (even as an quite advanced user! [^1]) in vast majority of cases much prefer simpler Quest-based StreetComplete instead of using Overlays.

While overlays are sometimes indeed very nice and useful, there are huge problems associated with them:
- even with recent improvements for easier (two-click) changing of overlay, it is still extremely taxing if one needs to use more than one overlay. Even in [SCEE](https://wiki.openstreetmap.org/wiki/SCEE) (which offers **one-click** changing of overlays) using Overlays is enormously less usable than using Quests when more than one overlay is wanted (which is very often, if not always[^2])
- number of overlays is growing, and will likely continue to do so. That would likely result in switching overlays even more complex in the future (i.e. certainly not much simpler)
- I know the default answer _"But you are only supposed to use one overlay, instead of switching them all the time, you are using them wrong"_, but I disagree. Even if one is fachidiot who only cares about single specific issue and nothing else (like I might be for bicycle-related infrastructure, for example), using one overlay simply does not cut it. 
  For example, if I care about suitability of the driving bicycle on the road, but that includes (at the very least) `surface overlay`, `smoothness overlay`, `track_type overlay`, `cycleway overlay` and probably a few more (`lit overlay`, `sidewalks overlay`). It is very usable and enjoyable using quests to solve that. But if the quest are deleted and I have to switch through several overlays every few dozen meters, it would be a nightmare (for hopefully obvious reasons)
- I find the idea that e.g. _"there are many users who care about smoothness but not surface"_ very unconvincing. 
  While merging several very related overlays (e.g. `surface`+`smoothness`+`tracktype`) into one gargangutan overlay would help with _"switching overlays"_ issue, it would significantly impact view-map usability (e.g. even if one combines colors+line shapes+outlines to represent multiple aspects, it would still be very hard to read -- I know, as I use such setup in OsmAnd to display `surface`+`smoothness`+`lit`+`access` *all at the same time* in some views)

---

**TL;DR**: Please don't remove smoothness quest. And generally, please let's not throw away Quest model (which made SC popular) in favor of (much less usable, IMHO) Overlay model.

[^1]:  less-advanced / beginner people that I know who use StreetComplete don't touch overlays at all.
[^2]:  In my experience, for way-based overlays, one basically always want more then one Overlay, even if mapping very specific things. (POI-based Overlays, on the other hands, might have use case for one-Overlay-only use; e.g. if one only care about addresses).

**Comment by rhhsm** [comment]
I agree with @mnalis : I mostly rely on quests to contribute with SC, and use overlays mostly to add information to the map that I can't add with quests. This includes surface, lighting and sidewalks of service roads (I think users will be interested in these for parking aisles, service roads that may be used by pedestrians, etc.), missing shops, and occasionally for adding street parking and addresses for buildings that should have one but that are of unclear exact status (single or multi family residences).
A new smoothness overlay will be useful to get an overview of an area and to correct mistakes, but should not replace the smoothness quest, I think.

**Comment by qugebert** [comment]
Am i allowed to this?

**Comment by mnalis** [comment]
> Am i allowed to this?

I don't think there was any opposition to writing the `Smoothness overlay` (to the contrary, it seems we'd all like that, just don't want Smothness Quest to go away as a result), and the issue is tagged as _"**new quest**  - accepted new quest proposal (if marked as blocked, it may require upstream work first)"_ and you have been assigned and nobody complained, so I would say you're welcome!

(It might be good idea to open _"Draft PR"_ mentioning it fixes this issue, so link is established and people can follow progress)

**Comment by mnalis** [comment]
> Am i allowed to this?

I don't think there was any opposition to writing the `Smoothness overlay` (to the contrary, it seems we'd all like that, just don't want Smothness Quest to go away as a result), and the issue is tagged as _"**new quest**  - accepted new quest proposal (if marked as blocked, it may require upstream work first)"_ and you have been assigned @qugebert  and nobody complained, so I would say you're welcome!

(It might be good idea to open _"Draft PR"_ mentioning it fixes this issue, so link is established and people can follow progress)

**Comment by mnalis** [comment]
> Am i allowed to this?

So, @qugebert, are you still interested in working on Smoothness overlay (on which you've been assigned on your request some time ago)?

**Comment by qugebert** [comment]
Argh sure, i just had some issue with a disappearing ok button on my last attempt and then somewhat lost sight about that.
I'll take another look at it next week and if I still can't get any further I'll at least get back to you with the current status, maybe someone has an idea what's wrong.

**Comment by westnordost** [comment]
The OK button is not shown if the form is not complete, IIRC

**Comment by qugebert** [comment]
Yeah i know, but i checked the isFormComplete etc functions, their returns etc. and it was fine.
I will have a look on it the next days.

**Comment by qugebert** [comment]
I'm on it, currently I have at least a “simple” working version, which only handles smoothness without prefixes. I'll test a bit this afternoon. Do we want to treat it like the surface overlay, that both values are queried for segregated foot/bike? This would then be my next step.

Can we talk about the color scheme here? Unfortunately, I have little feeling for design considerations. Currently I have chosen rather random colors. 

```
    /* TODO: Reconsider Colors */
    private val Smoothness.color get() = when (this) {
        Smoothness.EXCELLENT -> Color.BLUE
        Smoothness.GOOD -> Color.SKY
        Smoothness.INTERMEDIATE -> Color.CYAN
        Smoothness.BAD -> Color.AQUAMARINE
        Smoothness.VERY_BAD -> Color.TEAL
        Smoothness.HORRIBLE -> Color.ORANGE
        Smoothness.VERY_HORRIBLE -> Color.LIME
        Smoothness.IMPASSABLE -> Color.GRAY
        Smoothness.UNKNOWN -> Color.BLACK
    }
```

**Comment by rusty-snake** [comment]
Suggestions from me
- excellent/good/intermediate -> green shades
- bad/very_bad -> organe shades
- horrible/very_horrible -> brown shades

**Comment by FloEdelmann** [comment]
<details>
<summary>Potentially relevant for color considerations</summary>

https://github.com/streetcomplete/StreetComplete/blob/2f1ccecda85e467236bfb31ec5758ba1e64c67b5/app/src/main/java/de/westnordost/streetcomplete/overlays/surface/SurfaceOverlay.kt#L75-L99

https://github.com/streetcomplete/StreetComplete/blob/2f1ccecda85e467236bfb31ec5758ba1e64c67b5/app/src/main/java/de/westnordost/streetcomplete/overlays/Color.kt#L3-L24

</details>

**Comment by westnordost** [comment]
Thanks for the link, @FloEdelmann 

On the color palette, we have a sequence (bad -> good) here, so maybe we need a separate sequential color scheme, see 

https://personal.sron.nl/~pault/#sec:sequential

But we have a few more constraints: Red must stand out, plus we cannot variate the brightness a lot, because the overlay colors must remain vibrant to distinguish them from the background map. So, in this regard, I think the [current palette](https://davidmathlogic.com/colorblind/#%231A87E6-%232FACE8-%2330D4EE-%2310C1B8-%230DA082-%23F37D1E-%23EEBD0D-%23B6EF28) is pretty usable as a sequential palette, see

I'd actually compress the palette even more and assign a few values the same color, because otherwise the whole coloring would just consist of blue-ish to cyan-ish colors, as just about 10% of tagged ways is worse than "bad"

![Image](https://github.com/user-attachments/assets/ddb31e6a-f40f-4071-944a-5b7c3f42516c)

Like this [from top to bottom excellent, good, intermediate, bad, very bad, horrible + very horrible, impassable](https://davidmathlogic.com/colorblind/#%231A87E6-%232FACE8-%2330D4EE-%23B6EF28-%23EEBD0D-%23F37D1E-%230F0F0F)

By the way, since `smoothness` is a tag with well-defined values, any other values than the ones defined in the wiki can be treated as 
`INVALID` (i.e., red color), no need to reserve black for that.

**Comment by Helium314** [comment]
> By the way, since smoothness is a tag with well-defined values, any other values than the ones defined in the wiki can be treated as INVALID

Maybe `very_good` could be treated shown like good or excellent? It's not in the wiki list, but I find it useful considering the strict description for excellent. I quite often tagged otherwise excellent ways as good, because the "_As-new asphalt_" wasn't fulfilled (it was nice and smooth, but clearly not _as new_).

**Comment by matkoniecz** [comment]
> Maybe very_good could be treated shown like good or excellent? It's not in the wiki list

it really should be first discussed (and accepted) outside SC issue tracker

**Comment by qugebert** [comment]
To be on the safe side, I'll ask again here:
We treat smoothness like surface, so first of all we want a distinction in segregated footpaths and cycle paths here too, right? I.e. no “simple” survey of a common smoothness (as with lit etc.), right?
I wasn't sure whether the smoothness of segregated footpaths and cycle paths isn't usually identical (if they are being renovated, then both are being renovated at the same time), but on the other hand, it is particularly relevant for footpaths/cycle paths whether I can get through them easily with non-tired or poorly tired vehicles. 
Then I would probably adopt part of the logic of the SurfaceOverlayForm in a new superclass and also use it for the SmoothnessOverlayForm.
Is this procedure okay?

Translated with DeepL.com (free version)

**Comment by rhhsm** [comment]
Yes, it would be nice to set `smoothness` of segregated paths separately: if the `surface` is not the same, then `smoothness` is quite likely to be different as well. Newly renovated asphalt is usually `smoothness=excellent` while new paving stones are usually `smoothness=good` and sett is hardly ever better than `smoothness=intermediate`.

As for the colours to be used, I think it's best to use colour similar to the surface overlay, with `asphalt` and `excellent` being the same shade of blue, `paving_stones` the same colour as `good`, and `sett` = `intermediate`. For the worst `smoothness` values some of the colours for unpaved surfaces could be used (gravel, ground, etc.). If we start using it and it turns out to be confusing that some colours are the same as in the surface overlay, I suppose it's not too difficult to change them?

**Comment by qugebert** [comment]
Do you know where that sharp increase since 2020 comes from? https://taginfo.openstreetmap.org/keys/footway:surface#chronology
Could it be because SC has strongly pushed usage since then?
According to https://github.com/streetcomplete/StreetComplete/commits/master/app/src/main/java/de/westnordost/streetcomplete/quests/surface/AddFootwayPartSurface.kt?after=09fdc4e728ddb5191d11ae8d7cb4c8ba05f12dce+34, SC started asking at the end of 2019.
Before that we were under 5k here too

**Comment by westnordost** [comment]
Absolutely! Tagged with cycleway:smoothness etc.

**Comment by matkoniecz** [comment]
I guess that stub page at wiki is fine if anyone feels need to it.

(for example mentioning that StreetComplete contributed to much of its use is not self-evident even for people quite knowledgeable about OSM tagging)

**Comment by westnordost** [comment]
What's the status?

**Comment by rhhsm** [comment]
Curious too!

I'm also wondering if 
- all ways will be editable in the smoothness overlay (like in other overlays), or
- only ways that have a well-defined `surface=*` value will be editable (as it is now for the smoothness quest)?

Assuming that overlays are for a bit more advanced users, the second option may be acceptable without having images for each combination of `surface` and `smoothness` and may enable #5499 to be realised. 🤞

**Comment by westnordost** [comment]
> all ways will be editable in the smoothness overlay (like in other overlays), or

No.

What will enable #5499 to be realized is finding more and better example pictures, not skipping on SC ease-of-use.

**Comment by matkoniecz** [comment]
so lets take a look at what is a blocker here:

https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/quests/smoothness/AddRoadSmoothness.kt and https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/quests/smoothness/SmoothnessItem.kt

So currently `"asphalt", "concrete", "concrete:plates", "sett", "paving_stones", "compacted", "gravel", "fine_gravel"` are supported and ready for overlay

Note https://github.com/streetcomplete/StreetComplete/issues/5325#issuecomment-1772637797

> And in either case, these things should be contributed to the wiki first ( https://wiki.openstreetmap.org/wiki/Key:smoothness/Gallery ) and discussed, if necessary, in the forum, not in this issue tracker.

`unhewn_cobblestone` has some images - not sure whether there is something wrong with it, or is it viable to be added

`gravel` and `fine_gravel` miss horrible and very_horrible images on wiki while SC has something - what is going on here?

`dirt` / `earth` could get better images for basically any value

And looking at https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/osm/surface/SurfaceItem.kt

`concrete:lanes`, `mud`, `grass_paver`, `wood`, `woodchips`, `metal`/`metal_grid`, `pebbles`, `grass`, `sand`, `rock`, `ground` are missing altogether for now (`artificial_turf`, `clay`, `tartan` are special and not needed here)

https://github.com/streetcomplete/StreetComplete/issues/5499#issuecomment-2222373627 was offering some images

**Comment by mnalis** [comment]
> so lets take a look at what is a blocker here:

I'm wondering if @qugebert has already made some progress on those (could you @qugebert link current state of your Draft PR, if any?) or should we start looking from scratch?

**Comment by matkoniecz** [comment]
> or should we start looking from scratch?

Current blocker and probably 90% of work is finding or taking photos.

And for now code will be rather expanding `smoothness` quest and adding new images (and maybe descriptions?) to support new values. Only after we have decent coverage of values, maybe with exception of exotic stuff such as `surface=wood` or `metal` it would make sense to start coding overlay code itself

For photos: proposing, discussing them should take place on forums, probably in https://community.openstreetmap.org/t/how-to-map-key-smoothness/109860/60 (link to some new posts in the thread, feedback welcome)

**Comment by qugebert** [comment]
Argh, 
sorry for the late response, unfortunately a serious illness in the family at the end of March intervened and then I completely lost sight of this issue.

I was basically done with the overlay except for a few special cases that I need to look over again. I'll post a PR draft in the next days.. 

I've left the surfaces as they are for now, as we want to have a separate image for each surface-smoothness combination and not every smoothness value makes sense for every surface. But i dont think this should be a blocker for the overlay. I think we can extend the surfaces step by step.

**Comment by michaelblyons** [comment]
Is this blocked on the migration to Jetpack Compose?

**Comment by matkoniecz** [comment]
No.  See https://github.com/streetcomplete/StreetComplete/issues/5486#issuecomment-2850875173

> Current blocker and probably >95% of work is finding or taking photos.

> For photos: proposing, discussing them should take place on forums, probably in https://community.openstreetmap.org/t/how-to-map-key-smoothness/109860/60 (link to some new posts in the thread, feedback welcome)

If you can take photos (and share them on open license) or find missing ones (on open license) that would fill gaps in https://wiki.openstreetmap.org/wiki/Key:smoothness/Gallery then posting them in [this](https://community.openstreetmap.org/t/how-to-map-key-smoothness/109860) thread would help.

Or commenting/upvoting/downvoting posts with good/bad suggestions, especially ones later than  https://community.openstreetmap.org/t/how-to-map-key-smoothness/109860/98 (from https://wiki.openstreetmap.org/w/index.php?title=Key:smoothness/Gallery&action=history I see that I made edits based on content from several posts already)

------------

EDIT: thread is now processed by me up to https://community.openstreetmap.org/t/how-to-map-key-smoothness/109860/166

https://wiki.openstreetmap.org/wiki/Surface-smoothness_fényképes_táblázat may have content worth using

smoothness=good candidate - https://wiki.openstreetmap.org/wiki/File:Wood-2-Good-B.jpg

https://wiki.openstreetmap.org/wiki/File:Metal-3-Intermediate-B.jpg
intemediate candidate?

pull https://wiki.openstreetmap.org/wiki/Surface-smoothness_f%C3%A9nyk%C3%A9pes_t%C3%A1bl%C3%A1zat

https://wiki.openstreetmap.org/wiki/File:Wood-2-Good-A.jpg

https://wiki.openstreetmap.org/wiki/File:Ground-7-VeryHorrible-B.jpg

**Comment by matkoniecz** [comment]
Also, for specific image requests: anyone got low quality `surface=grass_paver` that they can take photo of?

Something that is hard to pass with regular car?

All I have seen and could make photos of and all found on Commons are well-passable by car.

EDIT: now asked on https://community.openstreetmap.org/t/anyone-got-low-quality-surface-grass-paver-that-they-can-take-photo-of/132380 and on Discord OpenStreetMap World

EDIT2: https://community.openstreetmap.org/t/anyone-got-low-quality-surface-grass-paver-that-they-can-take-photo-of/132380/ has a helpful answer but it cannot yet be used

**Comment by matkoniecz** [comment]
https://wiki.openstreetmap.org/wiki/Key:smoothness/Gallery now has images for `surface=wood`

I have not made attempt to review them or to check whether they are usable in StreetComplete. But maybe surface quest can now support one more `surface` value? (and get us closer to overlay)

or we can identify what exactly is missing and look for missing `wood` images

**Comment by matkoniecz** [comment]
BTW, perspective for photos ideally would be kind of like 

https://github.com/streetcomplete/StreetComplete/blob/5d5f8b05ab27b841aab5ca82438aa26c88f59bd5/app/src/androidMain/res/drawable-xxhdpi/surface_asphalt_very_bad.jpg
https://github.com/streetcomplete/StreetComplete/blob/5d5f8b05ab27b841aab5ca82438aa26c88f59bd5/app/src/androidMain/res/drawable-xxhdpi/surface_asphalt_intermediate.jpg
https://github.com/streetcomplete/StreetComplete/blob/5d5f8b05ab27b841aab5ca82438aa26c88f59bd5/app/src/androidMain/res/drawable-xxhdpi/surface_asphalt_good.jpg
https://github.com/streetcomplete/StreetComplete/blob/5d5f8b05ab27b841aab5ca82438aa26c88f59bd5/app/src/androidMain/res/drawable-xxhdpi/surface_asphalt_excellent.jpg

not sure which one of them is best

**Comment by matkoniecz** [comment]
[grass_paver](https://wiki.openstreetmap.org/wiki/Tag:surface%3Dgrass_paver) images were collected in https://wiki.openstreetmap.org/wiki/Key:smoothness/Gallery thanks to https://community.openstreetmap.org/t/anyone-got-low-quality-surface-grass-paver-that-they-can-take-photo-of/132380/

Not investigated yet whether they are usable for StreetComplete

If anyone is interested in this issue - feel free to try using them (and if they are insufficient - maybe better one can be obtained in the thread)

**Comment by westnordost** [comment]
I think grass paver doesn't need to be damaged to be classified `bad`. Big bumps through either holes (space between is not filled with earth/grass) or bumps (grass knoll is sprawling from the spaces between) should be enough to make them `bad`.

Your photo of `intermediate` grass paver is maybe not that representative? After all, there is no grass visible at all. 
I added some images to the forum thread.

I tend to favor either of those above for `intermediate`, a picture of "normal" grass paver for `bad` without big damages but visible holes (like https://commons.wikimedia.org/wiki/File:Am_Zwinger_Pirna_119632082.jpg) and https://wiki.openstreetmap.org/wiki/File:Uneven_grass_paver_3.jpg for `very_bad`.

How about the descriptions? I think it makes sense to not use the descriptions from paving stones 1:1 but also refer to the spacings in the grass pavers in the descriptions.

**Comment by matkoniecz** [comment]
> I think grass paver doesn't need to be damaged to be classified bad.

Yeah, regular grass paver may be fairly bad for city bikes, motorscooters mentioned in `intermediate`

Do you plan to post proposed change in https://community.openstreetmap.org/t/how-to-map-key-smoothness/109860/ ?
